comp/metadata/host/hostimpl/host.go

```diff
@@ -125,3 +135,34 @@ func (h *host) GetPayloadAsJSON(ctx context.Context) ([]byte, error) {
 func (h *host) fillFlare(fb flaretypes.FlareBuilder) error {
 	return fb.AddFileFromFunc(filepath.Join("metadata", "host.json"), func() ([]byte, error) { return h.GetPayloadAsJSON(context.Background()) })
 }
+
+func (h *host) writePayloadAsJSON(w http.ResponseWriter, _ *http.Request) {
+	jsonPayload, err := h.GetPayloadAsJSON(context.Background())
+	if err != nil {
+		apiutils.SetJSONError(w, h.log.Errorf("Unable to marshal v5 metadata payload: %s", err), 500)
+		return
+	}
+
+	scrubbed, err := scrubber.ScrubBytes(jsonPayload)
+	if err != nil {
+		apiutils.SetJSONError(w, h.log.Errorf("Unable to scrub metadata payload: %s", err), 500)
+		return
+	}
+	w.Write(scrubbed)
+}
+
+func (h *host) writeGohaiPayload(w http.ResponseWriter, _ *http.Request) {
+	payload := gohai.GetPayloadWithProcesses(pkgconfig.IsContainerized())
+	jsonPayload, err := json.MarshalIndent(payload, "", "  ")
+	if err != nil {
+		apiutils.SetJSONError(w, h.log.Errorf("Unable to marshal gohai metadata payload: %s", err), 500)
+		return
+	}
+
+	scrubbed, err := scrubber.ScrubBytes(jsonPayload)
+	if err != nil {
+		apiutils.SetJSONError(w, h.log.Errorf("Unable to scrub gohai metadata payload: %s", err), 500)
+		return
+	}
+	w.Write(scrubbed)
+}
```

Comment:
While having all endpoints registered inside components, we have to be mindful that sometimes not all endpoint belong to a component. 

In this case the `gohai` endpoint seems to not belong to the metadata host component as it is not fetching any internal of the `*host` struct.

For this particular case I to not move the endpoint from the apiimpl (for now). Once we have all of the component related endpoints extracted we will be able to better asses what to do with the remianing ones.

Reply:
gohai is part of the metadata/v5 payload and namespaced in the api `/metadata/`, I think it makes sense to keep it on the host metadata component?

Reply:
You are right! The gohai payload is part of the metadata v5 payload. 

I've look at how the host payload, and found that it is already part of the payload field

https://github.com/DataDog/datadog-agent/blob/main/comp/metadata/host/hostimpl/payload.go#L55-L62

```go
// getPayload returns the complete metadata payload as seen in Agent v5
func (h *host) getPayload(ctx context.Context) *Payload {
	p := &Payload{
		CommonPayload: *utils.GetCommonPayload(h.hostname, h.config),
		Payload:       *utils.GetPayload(ctx, h.config),
	}

	if r := h.resources.Get(); r != nil {
		p.ResourcesPayload = r["resources"]
	}

	if h.config.GetBool("enable_gohai") {
		gohaiPayload, err := gohai.GetPayloadAsString(pkgconfig.IsContainerized())
		if err != nil {
			h.log.Errorf("Could not serialize gohai payload: %s", err)
		} else {
			p.GohaiPayload = gohaiPayload
		}
	}
	return p
}
```

By looking at the code it looks like that gohai payload is collected if it enabled. but looking at the `writeGohaiPayload` HTTP handler we do not check that flag. I'm not sure if that is on purpose we should probably double check. I know the code was like that before, but it would be good to double check.

Also, with that in mind we could probably use the `payload := h.getPayload(ctx)` inside  `writeGohaiPayload` since it gathers the gohai payload as well. 


```go
func (h *host) writeGohaiPayload(w http.ResponseWriter, _ *http.Request) {
	payload := host.getPayload(context.Background()) 
	jsonPayload, err := json.MarshalIndent(payload.GohaiPayload, "", "  ")
	if err != nil {
		apiutils.SetJSONError(w, h.log.Errorf("Unable to marshal gohai metadata payload: %s", err), 500)
		return
	}

	scrubbed, err := scrubber.ScrubBytes(jsonPayload)
	if err != nil {
		apiutils.SetJSONError(w, h.log.Errorf("Unable to scrub gohai metadata payload: %s", err), 500)
		return
	}
	w.Write(scrubbed)
}
```

What do you think? @jeremy-hanna

Reply:
I think it's intentional that the endpoint does not check if it's enabled, that way the cli command `datadog-agent diagnose show-metadata gohai` will still return the gohai payload even if we aren't sending it to the backend

I think we shouldn't change the function - `host.getPayload` does more work (see timing screenshots) and formats it differently since `gohai.GetPayloadAsString` has slightly different behavior and does not include the processes which are nested under "resources" in the v5 payload

`curl /agent/metadata/v5` ~= 8s
<img width="150" alt="time curl v5" src="https://github.com/DataDog/datadog-agent/assets/16483405/9235c77d-673c-41f3-b650-0cdfb00e7d97">

`curl /agent/metadata/gohai` ~= 0.14s
<img width="208" alt="time curl gohai" src="https://github.com/DataDog/datadog-agent/assets/16483405/c8f1f940-9926-4ba7-a51c-55422b2376cc">